step/step.go

```diff
@@ -0,0 +1,255 @@
+
+	_, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to install XCTestHTMLReport: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ReportGenerator) Run(config Config) (Result, error) {
+	r.logger.Println()
+	r.logger.Infof("Collecting xcresult files")
+
+	patterns := []string{
+		fmt.Sprintf("%s/**/*.xcresult", config.TestDeployDir),
+	}
+	if 0 < len(config.XcresultPatterns) {
+		patterns = config.XcresultPatterns
+	}
+
+	paths, err := collectFilesWithPatterns(patterns)
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to find all xcresult files: %w", err)
+	}
+
+	if len(paths) == 0 {
+		r.logger.Printf("No files found.")
+
+		return Result{
+			HtmlReportDir: "",
+		}, nil
+	}
+
+	r.logger.Printf("List of files:")
+	for _, path := range paths {
+		r.logger.Printf("- %s", path)
+	}
+
+	rootDir, err := r.htmlReportsRootDir()
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to create test report directory: %w", err)
+	}
+
+	r.logger.Println()
+	r.logger.Infof("Generating reports")
+
+	for _, path := range paths {
+		if err := r.generateTestReport(rootDir, path); err != nil {
+			r.logger.Errorf("failed to generate test report (%s): %w", path, err)
+		}
+	}
+
+	r.logger.Println()
+	r.logger.Donef("Finished")
+
+	return Result{
+		HtmlReportDir: rootDir,
+	}, nil
+}
+
+func (r *ReportGenerator) Export(result Result) error {
+	return r.exporter.ExportOutput(htmlReportDirKey, result.HtmlReportDir)
+}
+
+func (r *ReportGenerator) generateTestReport(rootDir string, xcresultPath string) error {
+	baseName := strings.TrimSuffix(filepath.Base(xcresultPath), filepath.Ext(xcresultPath))
+	dirPath := filepath.Join(rootDir, baseName)
+	err := os.Mkdir(dirPath, 0755)
+	if err != nil {
+		if os.IsExist(err) {
+			r.logger.Warnf("Html report already exists for %s at %s", baseName, dirPath)
+			return nil
+		}
+		return err
+	}
+
+	r.logger.Printf("Generating report for: %s", baseName)
+	params := []string{"run", "bitrise-io/XCTestHTMLReport@speed-improvements", "--output", dirPath, xcresultPath}
+	cmd := r.commandFactory.Create("mint", params, nil)
+	_, err = cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to export html: %w", err)
+	}
+
+	if err := moveAssets(xcresultPath, dirPath); err != nil {
+		return fmt.Errorf("failed to move assets: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ReportGenerator) htmlReportsRootDir() (string, error) {
+	reportDir := r.envRepository.Get(htmlReportDirKey)
+	if reportDir == "" {
+		return os.MkdirTemp("", "html-reports")
+	}
+
+	exists, err := pathutil.NewPathChecker().IsDirExists(reportDir)
+	if err != nil {
+		return "", err
+	}
+	if !exists {
+		return "", fmt.Errorf("html report dir (%s) does not exist or is not a folder", reportDir)
+	}
+
+	return reportDir, nil
+}
+
+func collectFilesWithPatterns(patterns []string) ([]string, error) {
+	// Go does not have a set, so a map will help filter out duplicate results.
```

Comment:
In what scenario can it happen that there are duplicate file paths?

Reply:
You can enter multi glob patterns which will find the same files. For example:
```
/path/to/test.xcresult
/path/**/*.xcresult
```